README
The implementations in this project are done to learn new
algorithms. They are defintely not production quality and would
require refactoring and additional testing before I feel fully
confident about their correctness and robustness.


NEWS.txt
News
====

0.1
---

*Release date: UNRELEASED*

* In progress


tests/tests.py
from test_convex_hull import TestConvexHull
from test_convex_hulls import TestConvexHulls
from test_delaunay import TestDelaunay


tests/test_delaunay.py
import unittest
import itertools
import math
import numpy as np
from algbox.delaunay import Delaunay
from algbox.convex_hull import ConvexHull

class TestDelaunay(unittest.TestCase):
    def correctness_tests(self, n, reps):
        for i in range(reps):
            d = Delaunay(n=n, seed=i)
            d.compute_triangulation()
            self.assertTrue(d.validate_triangulation())
            self.assertTrue(self.validate_through_flip_check(d))

    @staticmethod
    def validate_through_flip_check(d):
        """Validates a Delaunay triangulation by verifying that no flips need
        to be done. This is here as an alternative to the validation implementation
        in the Delaunay class.

        Returns True if validation passes."""
        for tri_id1, triangle1 in d.triangles.items():
            for edge in itertools.combinations(triangle1, 2):
                # Get the other triangle sharing this edge, if one exists.
                if len(d.edge_mapping[edge]) == 2:
                    angle1 = d.get_angle(triangle1, list(set(triangle1)-set(edge))[0])
                    tri_id2 = list(d.edge_mapping[edge].difference([tri_id1]))[0]
                    triangle2 = d.triangles[tri_id2]
                    angle2 = d.get_angle(triangle2, list(set(triangle2)-set(edge))[0])

                    if angle1+angle2 > math.pi:
                        return False
        return True

    def test_correctness_n3(self):
        self.correctness_tests(3, 500)
            
    def test_correctness_n4(self):
        self.correctness_tests(4, 400)
            
    def test_correctness_n5(self):
        self.correctness_tests(5, 400)
            
    def test_correctness_n6(self):
        self.correctness_tests(6, 300)
            
    def test_correctness_n7(self):
        self.correctness_tests(7, 250)
            
    def test_correctness_n10(self):
        self.correctness_tests(10, 100)
            
    def test_correctness_n30(self):
        self.correctness_tests(30, 30)
            
    def test_validate(self):
        for i in range(100):
            d = Delaunay(n=15, seed=i)
            d.compute_triangulation()
            self.assertTrue(d.validate_triangulation())
            self.assertTrue(self.validate_through_flip_check(d))

    def test_validate_edge_mapping(self):
        """Verify the edge mapping is correct after triangulation."""
        n = 25
        d = Delaunay(n=n, seed=50)
        d.compute_triangulation()
        tri_edge_count = np.zeros((n, n), dtype=np.int32)
        edge_mapping_count = tri_edge_count.copy()
        for edge, emap in d.edge_mapping.items():
            i = edge[0]
            j = edge[1]
            self.assertTrue(i < j)
            edge_mapping_count[i, j] += len(d.edge_mapping[edge])

        for triangle in d.triangles.values():
            for edge in itertools.combinations(triangle, 2):
                i = edge[0]
                j = edge[1]
                self.assertTrue(i < j)
                tri_edge_count[i, j] += 1

        self.assertTrue(np.all(tri_edge_count == edge_mapping_count))
        self.assertEquals(np.max(tri_edge_count), 2)

    @staticmethod
    def shared_edge_angles_not_oblique(d, emap):
        """Returns True if the angles of the shared edge between the
        two triangles in emap are not oblique. Put another way, it
        is True if the convex hull of the two triangles contains all
        four unique points in the two triangles."""
        id1, id2 = emap
        tri_point_ids = set(d.triangles[id1]).union(set(d.triangles[id2]))
        triangles_points = np.array([d.points[id] for id in list(tri_point_ids)])
        chull = ConvexHull(triangles_points)
        return len(chull) == 4

    def test_validation_bad_triangulation(self):
        """Test the validation method by taking a correct triangulation,
        flip a pair of triangles, verify the new triangulation is bad,
        restore the triangulation, and move to a new pair of triangles.

        Only triangle pairs that have a convex hull containing their four
        unique points are included in the flipping."""
        n = 25
        d = Delaunay(n=n, seed=50)
        d.compute_triangulation()

        self.assertTrue(d.validate_triangulation())
        self.assertTrue(self.validate_through_flip_check(d))

        # Now break the triangulation.
        for emap in d.edge_mapping.values():
            if len(emap) == 2 and self.shared_edge_angles_not_oblique(d, emap):
                id1, id2 = emap
                d.flip(id1, id2)
                self.assertFalse(d.validate_triangulation())
                self.assertFalse(self.validate_through_flip_check(d))
                d.flip(id1, id2)
                self.assertTrue(d.validate_triangulation())
                self.assertTrue(self.validate_through_flip_check(d))


tests/test_convex_hull.py
from random import random, seed
from pylab import *

import unittest
import itertools
import numpy as np
from algbox.convex_hull import ConvexHull

class TestConvexHull(unittest.TestCase):
    def test_hull_n3(self):
        """Test the convex hull algorithm with three point input."""
        points = np.array([[1, 1], [0, 0], [-1, 1]])
        ch = ConvexHull(points)
        self.assertTrue(np.all([1, 0, 2] == ch.hull))

        points = np.array([[2, 1], [0, 0], [1, 1]])
        ch = ConvexHull(points)
        self.assertTrue(np.all([1, 0, 2] == ch.hull))

    def test_hull_n4(self):
        """Test the convex hull algorithm with four point input."""
        # All point are in the hull.
        points = np.array([[1, 1], [0, 0], [-1, 1], [0.8, 0.5]])
        ch = ConvexHull(points)
        self.assertTrue(np.all([1, 3, 0, 2] == ch.hull))

        # One of the points is not in the hull.
        points = np.array([[1, 1], [0, 0], [-1, 1], [0.2, 0.5]])
        ch = ConvexHull(points)
        self.assertTrue(np.all([1, 0, 2] == ch.hull))

    def test_hull_n25(self):
        n = 25
        seed(1)
        points = np.array([random() for i in range(2*n)]).reshape((n,2))
        ch = ConvexHull(points)
        ## for i in range(n):
        ##     annotate(i, points[i])
        ## show()
        self.assertTrue(np.all([13, 23, 16, 20, 12, 10, 14, 19, 1] == ch.hull))

    def test_is_right_turn(self):
        """Test the is_right_turn method."""
        points = np.array([[1, 1], [0, 0], [-1, 1], [0.8, 0.5]])
        ch = ConvexHull(points)
        ch.convex_hull = [1, 2]
        self.assertTrue(ch.is_right_turn(0))
        self.assertTrue(ch.is_right_turn(3))

        ch.convex_hull = [0, 1]
        self.assertTrue(ch.is_right_turn(2))
        self.assertFalse(ch.is_right_turn(3))

        ch.convex_hull = [1, 0]
        self.assertFalse(ch.is_right_turn(2))
        self.assertTrue(ch.is_right_turn(3))

    def test_line_cosine(self):
        """Test the line_cosine method."""
        points = np.array([[1, 1], [0, 0], [-1, 1], [0.8, 0.5]])
        ch = ConvexHull(points)
        self.assertTrue(np.allclose(ch.line_cosine(0, 1), math.cos(5*math.pi/4)))
        self.assertTrue(np.allclose(ch.line_cosine(0, 2), math.cos(math.pi)))
        self.assertTrue(np.allclose(ch.line_cosine(1, 3), math.cos(0.5585993153435626)))

    def test_point_distance(self):
        """Test the point_distance method."""
        p1 = np.array([0, 0])
        p2 = np.array([0, 7])
        self.assertEquals(ConvexHull.point_distance(p1, p2), 7)

        p1 = np.array([3.5, 7.4])
        p2 = np.array([-6, -9.2])
        self.assertEquals(ConvexHull.point_distance(p1, p2), 19.126160095534075)

    def test_contains(self):
        """Test to verify that statements like

            >>> ch = ConvexHull(points)
            >>> 3 in ch

        behave as expected."""
        n = 25
        seed(1)
        points = np.array([random() for i in range(2*n)]).reshape((n,2))
        ch = ConvexHull(points)
        expected = [13, 23, 16, 20, 12, 10, 14, 19, 1]
        self.assertTrue(np.all(expected == ch.hull))
        for exp in expected:
            self.assertTrue(exp in ch)
        self.assertFalse(15 in ch)
        self.assertFalse(25 in ch)

    def test_len(self):
        """Test to verify statements like

            >>> ch = ConvexHull(points)
            >>> len(ch)

        behave as expected."""
        n = 25
        seed(1)
        points = np.array([random() for i in range(2*n)]).reshape((n,2))
        ch = ConvexHull(points)
        self.assertTrue(np.all([13, 23, 16, 20, 12, 10, 14, 19, 1] == ch.hull))
        self.assertEquals(len(ch), 9)

    def test_getitem(self):
        """Test to verify statements like

            >>> ch = ConvexHull(points)
            >>> ch[0]

        behave as expected."""
        n = 25
        seed(1)
        points = np.array([random() for i in range(2*n)]).reshape((n,2))
        ch = ConvexHull(points)
        expected = [13, 23, 16, 20, 12, 10, 14, 19, 1]
        self.assertTrue(np.all(expected == ch.hull))
        for i in range(len(ch)):
            self.assertEquals(ch[i], expected[i])
        self.assertRaises(IndexError, ch.__getitem__, len(ch))


src/algbox/convex_hull.py
import numpy as np
import math

class ConvexHull(object):
    def __init__(self, points):
        self.points = points
        self.compute_hull()

    def __contains__(self, val):
        return val in self.convex_hull

    def __getitem__(self, index):
        return self.convex_hull[index]

    def __len__(self):
        return len(self.convex_hull)

    @property
    def x(self):
        return self.points[:,0]

    @property
    def y(self):
        return self.points[:,1]

    @property
    def hull(self):
        return self.convex_hull

    def compute_hull(self):
        """Compute the convex hull of the point field using Graham scan."""
        # Start with the hull containing the lowest point in the point field.
        self.convex_hull = [np.argmin(self.points[:,1], 0)]

        # Get the cosines of the angles of all lines between the "lowest" point
        # and all other points.
        ids_and_cos = np.array([[i,self.line_cosine(self.convex_hull[0], i)] for i in range(len(self.points)) if i != self.convex_hull[0]])

        # Get a sorted list of point ids for points furthest to right (by
        # angle) to furthest left, from the perspective of the lowest point.
        ids = ids_and_cos[np.argsort(ids_and_cos[:,1], 0)][::-1][:,0]

        # Now go through the sorted list
        self.convex_hull.append(ids[0])
        for pid in ids[1:]:
            while self.is_right_turn(pid):
                self.convex_hull.pop()
            self.convex_hull.append(pid)

        # Make sure the last point added does not introduce a right turn
        # to our first point.
        if self.is_right_turn(self.convex_hull[0]):
            self.convex_hull.pop()

        return np.array(self.convex_hull, dtype=np.int)

    def is_right_turn(self, id3):
        """Determines if point indicated by id is a "right turn" from the
        line made by the last two points in the convex hull."""
        id1 = self.convex_hull[-2]
        id2 = self.convex_hull[-1]
        cross_prod = (self.x[id2] - self.x[id1])*(self.y[id3] - self.y[id1]) - \
                     (self.y[id2] - self.y[id1])*(self.x[id3] - self.x[id1])
        return cross_prod < 0

    def line_cosine(self, id1, id2):
        """Computes the cosine of the angle between the x-axis and the line
        between the points referred to by id1 and id2."""
        return (self.x[id2] - self.x[id1]) / self.point_distance(self.points[id1],
                                                                 self.points[id2])

    @staticmethod
    def point_distance(p1, p2):
        """Computes the distance between two points."""
        return math.sqrt(math.pow(p1[0] - p2[0], 2) + math.pow(p1[1] - p2[1], 2))


src/algbox/convex_hulls.py
import argparse
import numpy as np
from random import random, seed
import sys
from convex_hull import ConvexHull

from pylab import axis, matplotlib, plot, Polygon, show
import matplotlib.pyplot as plt
from matplotlib.collections import PatchCollection
import matplotlib.patches as mpatches

class ConvexHulls(object):
    def __init__(self, points, max_hulls=1, plot_points=True):
        self.points = points
        self.plot_points = plot_points
        self.pids = np.arange(points.shape[0])
        self.convex_hulls = []
        interior_points = np.ones((self.points.shape[0]), dtype=np.bool)
        interior_points[:] = True

        while np.sum(interior_points) > 2 and len(self.convex_hulls) < max_hulls:
            ch = ConvexHull(points[interior_points])
            ipids = self.pids[interior_points]
            self.convex_hulls.append([ipids[id] for id in ch.convex_hull])
            for pid in ch.convex_hull:
                interior_points[ipids[pid]] = False

    def __len__(self):
        return len(self.convex_hulls)

    def __getitem__(self, index):
        return self.convex_hulls.__getitem__(index)

    @property
    def x(self):
        return self.points[:,0]

    @property
    def y(self):
        return self.points[:,1]

    def plot(self):
        """Plot the point field and convex hulls."""
        fig = plt.figure(figsize=(8,8))
        ax = plt.axes([0,0,1,1])
        if self.plot_points:
            plot(self.x, self.y, 'ok')  # plot points

        # Draw hulls.
        patches = []
        for hull in self.convex_hulls:
            hull_points = np.array([self.points[pid] for pid in hull])
            patches.append(Polygon(hull_points))

        colors = 100*np.random.rand(len(patches))
        collection = PatchCollection(patches, cmap=matplotlib.cm.jet, alpha=0.1, lw=2, linestyle='solid')
        collection.set_array(np.array(colors))
        ax.add_collection(collection)

        axis((0, 1, 0, 1))

def parse_input():
    parser = argparse.ArgumentParser(description='Compute convex hull of a random point field.')
    parser.add_argument('-n', '--npoints', dest='npoints', action='store',
                        default=20,
                        help='number of points in the randomly-generated point field (default: %(default)s)')
    parser.add_argument('--no-plot-points', dest='plot_points', action='store_false',
                        help='do not plot points')
    parser.add_argument('--num-hulls', dest='nhulls', action='store', default=1,
                        help="set the number of hulls to compute; set to 'all' to compute all possible hulls (default: %(default)s)")
    parser.add_argument('-s', '--seed', dest='seed', action='store', default=None,
                        help='set the random seed (allows for reproducibility)')
    return parser.parse_args()

def gen_random_points(n):
    """Generate random points in [0,1] x [0,1]."""
    return np.array([random() for i in range(2*n)]).reshape((n,2))

def convex_hulls():
    args = parse_input()

    if args.seed is not None:
        seed(int(args.seed))
    points = gen_random_points(int(args.npoints))
    if args.nhulls == 'all':
        nhulls = sys.maxint
    else:
        nhulls = int(args.nhulls)
    ch = ConvexHulls(points, nhulls, args.plot_points)
    ch.plot()
    show()


src/algbox/delaunay.py
import argparse
from pylab import *
import numpy as np
import random
import itertools
import math
from algbox.convex_hull import ConvexHull

import matplotlib.pyplot as plt
from matplotlib.collections import PatchCollection
import matplotlib.patches as mpatches

class Delaunay(object):
    def __init__(self, n, seed=None, summary=False, verbose=False):
        if seed is not None:
            random.seed(seed)
        self.points = self.gen_random_points(n)
        self.summary = summary
        if summary:
            self.ntriangles = 0
            self.nadds = 0
            self.nremoves = 0
            self.nflips = 0
        self.verbose = verbose

    def gen_random_points(self, n):
        """Generate random points in [0,1] x [0,1]."""
        return np.array([random.random() for i in range(2*n)]).reshape((n,2))

    @property
    def x(self):
        return self.points[:,0]

    @property
    def y(self):
        return self.points[:,1]

    def compute_triangulation(self):
        self.convex_hull = ConvexHull(self.points)

        self.triangle_id = 0
        self.triangles = {}
        self.edge_mapping = {}

        self.initial_triangles_from_hull()

        # Add remaining points incrementally.
        remaining_points = [i for i in range(len(self.points))
                            if i not in self.convex_hull]

        for p in remaining_points:
            self._accounting('add_point', p)
            in_triangle = self.point_in_triangle(p)
            self.split_triangle(in_triangle, p)

    def split_triangle(self, tri_id, point_id):
        """Split the triangle with ID tri_id into three triangles using the
        given point_id."""
        parent_triangle = self.triangles[tri_id]
        self._accounting('split', parent_triangle)
        if self.summary:
            self.nremoves += 1
            self.ntriangles -= 1
        self.remove_triangle(tri_id)
        for a,b in itertools.combinations(parent_triangle, 2):
            new_triangle = sort([a, b, point_id])
            self._accounting('add_triangle', new_triangle)
            self.add_triangle(new_triangle)
            for key in itertools.combinations(new_triangle, 2):
                if self.check_and_flip(key):
                    break

    def point_in_triangle(self, p):
        """Determine the triangle that p is inside by searching for
        the triangle ABC for which the sum of the angles ApB, BpC,
        and CpA is 2*pi."""
        for id,triangle in self.triangles.iteritems():
            angle_sum = 0
            for a,b in itertools.combinations(triangle, 2):
                temp_triangle = [a, b, p]
                angle_sum += self.get_angle(temp_triangle, p)
            if np.allclose([angle_sum], [2*math.pi], 1e-10, 0):
                return id

    def add_triangle(self, corners):
        """Add new triangle with given corners."""
        tri_id = self.triangle_id
        self.triangle_id += 1
        self.triangles[tri_id] = corners
        for key in itertools.combinations(corners, 2):
            if self.edge_mapping.has_key(key):
                self.edge_mapping[key].add(tri_id)
            else:
                self.edge_mapping[key] = set([tri_id])
        return tri_id

    def remove_triangle(self, id):
        """Remove the triangle with the given ID."""
        corners = self.triangles[id]
        for key in itertools.combinations(corners, 2):
            self.edge_mapping[key].difference_update([id])
            if len(self.edge_mapping[key]) == 0:
                del(self.edge_mapping[key])
        del(self.triangles[id])

    def _accounting(self, event, data=None):
        """This method is used to print and record information about the
        triangulation. It only has effect when the user requests verbose
        output or the summary."""
        if not self.summary and not self.verbose:
            return
        if event == 'add_point':
            if self.verbose:
                print 'Adding point {}'.format(data)
        elif event == 'add_triangle':
            if self.verbose:
                print '    Adding triangle {}'.format(data)
            if self.summary:
                self.nadds += 1
                self.ntriangles +=1
        elif event == 'build_hull':
            if self.verbose:
                print 'Building convex hull... {}'.format(data)
        elif event == 'create_initial':
            if self.verbose:
                print 'Creating initial triangle {}'.format(data)
            if self.summary:
                self.nadds += 1
                self.ntriangles += 1
        elif event == 'flip':
            if self.verbose:
                print '    Flipping triangles {} and {}'.format(data[0], data[1])
            if self.summary:
                self.nflips += 1
        elif event == 'select_interior':
            if self.verbose:
                print 'Selecting starting interior point... {}'.format(data)
        elif event == 'split':
            if self.verbose:
                print 'Splitting triangle'
                print '    Removing triangle {}'.format(data)
        else:
            raise ValueError('unknown event')

    def initial_triangles_from_hull(self):
        # Triangulate the convex hull.
        start_point = self.convex_hull[0]
        shifted_hull = np.concatenate((self.convex_hull[2:], [self.convex_hull[1]]))
        for h1,h2 in zip(self.convex_hull[1:], shifted_hull)[:-1]:
            tri_edges = sort([h1, h2, start_point])
            self._accounting('create_initial', tri_edges)
            self.add_triangle(tri_edges)
            for key in itertools.combinations(tri_edges, 2):
                if self.check_and_flip(key):
                    break

    def check_and_flip(self, shared_edge):
        """Check an adjacent pair of triangles and flip if necessary."""
        if len(self.edge_mapping[shared_edge]) == 2:
            id1, id2 = list(self.edge_mapping[shared_edge])
            tri1 = self.triangles[id1]
            angle1 = self.get_angle(tri1, list(set(tri1)-set(shared_edge))[0])
            tri2 = self.triangles[id2]
            angle2 = self.get_angle(tri2, list(set(tri2)-set(shared_edge))[0])

            if angle1+angle2 > math.pi:
                self._accounting('flip', [tri1, tri2])
                self.flip(id1, id2)

                # Check flipped triangles for the Delaunay condition and
                # flip as necessary.
                for key in itertools.combinations(self.triangles[id1], 2):
                    if self.check_and_flip(key):
                        break

                for key in itertools.combinations(self.triangles[id2], 2):
                    if self.check_and_flip(key):
                        break

                return True
        return False

    def flip(self, id1, id2):
        """Flip the common edge between two adjacent triangles.

        This should only be called for pairs of triangles where all four
        points are in the convex hull formed by the triangles' points.
        This constraint is not an issue for calls from method
        check_and_flip because triangle pairs where only three points
        are in the convex hull do not violate the opposite angle
        condition (i.e., the sum of the triangle corner angles opposite
        the common edge are guaranteed to be less then 180 degrees
        when only three points are in the convex hull)."""
        tri1 = self.triangles[id1]
        tri2 = self.triangles[id2]
        shared_edge = sort(list(set(tri1).intersection(tri2)))
        new_edge = sort(list(set(np.concatenate((tri1, tri2))).difference(shared_edge)))

        new_tri1 = sort(np.concatenate((new_edge, [shared_edge[0]])))
        new_tri2 = sort(np.concatenate((new_edge, [shared_edge[1]])))

        self.edge_mapping[tuple(new_edge)] = self.edge_mapping[tuple(shared_edge)]
        del(self.edge_mapping[tuple(shared_edge)])

        self.triangles[id1] = new_tri1
        self.triangles[id2] = new_tri2

        # Update edge_mapping, replacing triangle 2 with triangle 1
        # (and vice versa) for edges that were transferred between
        # triangles.
        for key in itertools.combinations(new_tri1, 2):
            if key != tuple(new_edge):
                self.edge_mapping[key].difference_update([id2])
                self.edge_mapping[key].add(id1)

        for key in itertools.combinations(new_tri2, 2):
            if key != tuple(new_edge):
                self.edge_mapping[key].difference_update([id1])
                self.edge_mapping[key].add(id2)

    def get_angle(self, triangle, corner):
        """Find the angle of the given corner of the given triangle."""
        non_corner = list(set(triangle) - set([corner]))
        # a is the edge opposite the corner
        a = self.point_distance(self.points[non_corner[0]], self.points[non_corner[1]])

        b = self.point_distance(self.points[corner], self.points[non_corner[0]])
        c = self.point_distance(self.points[corner], self.points[non_corner[1]])

        return math.acos((math.pow(b, 2) + math.pow(c, 2) - math.pow(a, 2)) / (2*b*c))

    @staticmethod
    def point_distance(p1, p2):
        """Computes the distance between two points."""
        return math.sqrt(math.pow(p1[0] - p2[0], 2) + math.pow(p1[1] - p2[1], 2))

    def plot(self):
        fig = plt.figure(figsize=(8,8))
        ax = plt.axes([0,0,1,1])        
        scatter(self.x, self.y)

        # Draw triangles.
        patches = []
        for triangle in self.triangles.itervalues():
            patches.append(self.draw_circumcircle(triangle))
            for i1,i2 in itertools.combinations(range(3), 2):
                p1 = triangle[i1]
                p2 = triangle[i2]
                plot([self.x[p1], self.x[p2]], [self.y[p1], self.y[p2]], 'k')

        colors = 100*np.random.rand(len(patches))
        collection = PatchCollection(patches, cmap=matplotlib.cm.jet, alpha=0.1)
        collection.set_array(np.array(colors))
        ax.add_collection(collection)

        axis((0, 1, 0, 1))

    def draw_circumcircle(self, triangle):
        [center_x, center_y], radius = self.get_circumcircle(triangle)
        return mpatches.Circle([center_x, center_y], radius, ec="none")

    def get_circumcircle(self, triangle):
        """Return the center and radius of the given triangle's
        circumcircle."""
        # Get the midpoints of two of the triangle's edges.
        mp1x,mp1y = self.get_midpoint(triangle[0], triangle[1])
        mp1_slope = -1./self.get_slope(triangle[0], triangle[1])
        mp1_intercept = self.get_intercept(mp1_slope, mp1x, mp1y)

        mp2x,mp2y = self.get_midpoint(triangle[1], triangle[2])
        mp2_slope = -1./self.get_slope(triangle[1], triangle[2])
        mp2_intercept = self.get_intercept(mp2_slope, mp2x, mp2y)

        # Compute intersection of the midpoint orthogonal lines.
        center_x = (mp2_intercept - mp1_intercept) / (mp1_slope - mp2_slope)
        center_y = mp1y + mp1_slope * (center_x - mp1x)
        radius = self.point_distance([center_x, center_y], self.points[triangle[0]])

        return [center_x, center_y], radius

    def get_intercept(self, slope, x, y):
        return y - slope*x

    def get_slope(self, id1, id2):
        rise = self.y[id1] - self.y[id2]
        run = self.x[id1] - self.x[id2]
        return rise/run

    def get_midpoint(self, id1, id2):
        mpx = (self.x[id1] + self.x[id2]) / 2
        mpy = (self.y[id1] + self.y[id2]) / 2
        return mpx,mpy

    def validate_triangulation(self):
        """Go through the triangles and validate that the triangulation
        is Delaunay."""
        for tri_id, triangle in self.triangles.iteritems():
            [center_x, center_y], radius = self.get_circumcircle(triangle)
            outside_point_ids = [pid for pid in range(len(self.points))
                                 if pid not in triangle]
            for pid in outside_point_ids:
                dist = self.point_distance([center_x, center_y], self.points[pid])
                if dist < radius:
                    return False
        return True

def parse_input():
    parser = argparse.ArgumentParser(description='Compute Delaunay triangulation on a random point field.')
    parser.add_argument('-n', '--npoints', dest='npoints', action='store',
                        default=20,
                        help='number of points in the randomly-generated point field (default: %(default)s)')
    parser.add_argument('--no-plot', dest='plot', action='store_false',
                        default=True, help='suppress plotting the triangulation')
    parser.add_argument('-s', '--seed', dest='seed', action='store', default=None,
                        help='set the random seed (allows for reproducibility)')
    parser.add_argument('--summary', dest='summary', action='store_true',
                        default=False,
                        help='compile triangulation summary information')
    parser.add_argument('-v', '--verbose', dest='verbose', action='store_true',
                        default=False,
                        help='print information as the triangulation proceeds')
    parser.add_argument('--validate', dest='validate', action='store_true',
                        default=False,
                        help='validate the computed triangulation is Delaunay')
    return parser.parse_args()

def delaunay():
    args = parse_input()

    if args.seed is not None:
        d = Delaunay(int(args.npoints), seed=int(args.seed), summary=args.summary,
                     verbose=args.verbose)
    else:
        d = Delaunay(int(args.npoints), verbose=args.verbose, summary=args.summary)
    d.compute_triangulation()
    if args.summary:
        summary_vals = [str(v) for v in [d.points.shape[0], d.ntriangles,
                                         d.nadds, d.nremoves, d.nflips]]
        max_val_len = max([len(v) for v in summary_vals])
        summary_text = ['points', 'triangles', 'triangles added',
                        'triangles removed', 'triangle flips']
        min_text_len = min([len(t) for t in summary_text])
        print '\n============= Summary ============='
        for val, text in zip(summary_vals, summary_text):
            print 'Number of {} {} {}{}'.format(text,
                                                '.'*(14+min_text_len-len(text)),
                                                ' '*(max_val_len-len(val)),
                                                val)
    if args.validate:
        if d.validate_triangulation():
            print "Validation: is Delaunay!"
        else:
            print "Validation: is not Delaunay!"
    if args.plot:
        d.plot()
        show()
